### ISSUE TITLE
Importing anki cards to org file

### ISSUE THREAD

**Issue by RoseWebb** [eyeinsky/org-anki#32]
Is there anyway to import the existing cards from a specifc deck to an org file? perhaps it can be added?

**Comment by eyeinsky** [comment]
Not at the moment. For you what would your general plan be in doing this: would you keep Anki app/web as the source of truth and would like to sync in the future, or would you like to do this just once?

In general I think this future is difficult to do well, because all features that Anki support would need their own "encoding" into org format. So when you sync from Anki to org, then edit, then sync back from org to anki it would need to look the same. org-anki at the moment converts org's syntax to html using org-mode's own tools, but there doesn't appear to be an official way to go the other way (html to org).

**Comment by RoseWebb** [comment]
I was planning to do this once and then continue using the org file as the main source. As you mentioned, I could not find existing solution to export Anki cards to org file. Therefore, I made a python script to use AnkiConnect and for writing the org file using OrgFormat. The html to org conversion is handled by pandoc. 

Now I can use your tool to sync to Anki.

**Comment by eyeinsky** [comment]
Right, perhaps there could be a command to pull a deck into a file for one-off purposes. (If you don't mind, maybe you could paste the script you used here?)

**Comment by RoseWebb** [comment]
Yeah that would be ideal. This is just bunch of snippets hacked together and its very slow. There is lots of room for improvement.
 
I also have some Cloze Overlapping type of notes in my Anki, I will have to figure out a way to sync them. For now I just have them in the org file, but not synced to Anki.

```import json
import orgformat
import urllib.request
import pypandoc


#---------------------------------------------#
# config
query = "deck:DeckName"
OrgFile = 'MyNotes.org'
#---------------------------------------------#

def request(action, **params):
    return {'action': action, 'params': params, 'version': 6}


def invoke(action, **params):
    requestJson = json.dumps(request(action, **params)).encode('utf-8')
    response = json.load(
        urllib.request.urlopen(
            urllib.request.Request('http://localhost:8765', requestJson)))
    if len(response) != 2:
        raise Exception('response has an unexpected number of fields')
    if 'error' not in response:
        raise Exception('response is missing required error field')
    if 'result' not in response:
        raise Exception('response is missing required result field')
    if response['error'] is not None:
        raise Exception(response['error'])
    return response['result']

OrgFormat = orgformat.OrgFormat
with open(OrgFile, 'w', encoding='utf-8') as f:
    NotesID = invoke("findNotes", query=query)
    for ID in NotesID:
        noteInfo = invoke("notesInfo", notes=[ID])[0]
        if noteInfo['modelName'].lower() == 'Cloze (overlapping)'.lower():
            Heading = pypandoc.convert_text(noteInfo['fields']['Title']['value'].replace("\n", " "),'org', format='html')
            Text = pypandoc.convert_text(noteInfo['fields']['Original']['value'].replace("<br>", ''), 'org', format='html')
            f.write(OrgFormat.generate_heading(level = 1, title = Heading.replace("\n"," "), properties =[("ANKI_NOTE_ID", str(ID))], section=Text))
        elif noteInfo['modelName'].lower() == 'Basic'.lower():
            Heading = pypandoc.convert_text(noteInfo['fields']['Front']['value'].replace("\n",' '), 'org', format='html')
            f.write(OrgFormat.generate_heading(level = 1, title = Heading.replace("\n"," "), properties =[("ANKI_NOTE_ID", str(ID))], section=Text))
            Text = pypandoc.convert_text(noteInfo['fields']['Back']['value'].replace("<br>", ''), 'org', format='html')
        elif noteInfo['modelName'].lower() == 'Cloze'.lower():
            Heading = pypandoc.convert_text(noteInfo['fields']['Text']['value'].replace("\n", " "), 'org', format='html')
            Text = pypandoc.convert_text(noteInfo['fields']['Back Extra']['value'].replace("<br>", ''), 'org', format='html')
            f.write(OrgFormat.generate_heading(level = 1, title = Heading.replace("\n"," "), properties =[("ANKI_NOTE_ID", str(ID))], section=Text))
        else:
            print(noteInfo['modelName'])``

**Comment by eyeinsky** [comment]
Thanks!

What do you mean by "Cloze Overlapping"? There is Cloze support already, but you may mean something else.

**Comment by RoseWebb** [comment]
Cloze Overlapping is a Special type of Cloze notes. They are not default in Anki, but added by an add-on [check their github for more info](https://github.com/glutanimate/cloze-overlapper)

**Comment by gsingh93** [comment]
I'd also find this useful, as I sometimes add notes to Anki through other means on mobile, and I'd like to get those new cards into Emacs as the source of truth. That being said it's not super high priority because I can just export the new notes to a file and use Python or regex to get them in the right format, although this takes some extra time.

**Comment by eyeinsky** [comment]
I've made some progress here (in the `import-deck` branch, perhaps I should just merge to master if it's easier to test that way -- let me know!).

Running `org-anki-import-deck` will prompt you for a deck name, open a new buffer and then pull the entire thing into that as top level org entries, with IDs (so updates work), adds note type to the entry if it's not Basic, and adds deck name file-globally.

It adds everything to the current buffer at the point where cursor is, so opening a new file and running the import there makes the most sense.

I didn't include running html-to-org conversion yet, but running the following pandoc command worked: `cat import.org | pandoc --from=html --to=org --wrap=preserve`. Perhaps it's better to integrate it though -- to run the command on each title/body separately instead of the whole file?

**Comment by hwiorn** [comment]
> What do you mean by "Cloze Overlapping"? There is Cloze support already, but you may mean something else.

Cloze overlapping is an overlapping text(or nested cloze deletion) technique of Anki. [Anki doesn't support overlapping clozes directly. The recommended way is that makes new text fields for every cloze.](https://docs.ankiweb.net/editing.html#cloze-deletion)

The concept is a cloze text inside main cloze fields. Then the clozes can be shown differently.
[It can use to overcome leeches.](https://eshapard.github.io/anki/the-power-of-making-new-cards-on-the-fly-in-anki.html) But a lot of work is needed to make new cards. The [LPCG](https://github.com/sobjornstad/AnkiLPCG) or the [cloze overlapper](https://github.com/glutanimate/cloze-overlapper) which is [inspired by the LPCG](https://github.com/glutanimate/cloze-overlapper#license-and-credits) make this easier.

I used this for memorizing speech and lyrics a lot. I am interested in that this can be implemented in org-anki.

**Comment by ghost** [comment]
Quick question, @eyeinsky, are you against an option to use subheadings as in anki-editor.el? Some types of cards would really benefit from having more than one line on the front page and overlapping cloze is one of them.

**Comment by eyeinsky** [comment]
Could you clarify the benefit? If there is an overlapping cloze like this `{{c1::Canberra was {{c2::founded}}}} in 1913` then already now you could create the following org headings:
```
* {{c1::Canberra was founded}} in 1913
* Canberra was {{c2::founded}} in 1913
```

And you can group the cards under a higher-level heading:
```
* Canberra was founded in 1913
** {{c1::Canberra was founded}} in 1913
** Canberra was {{c2::founded}} in 1913
```

And then run `org-anki-update-all` to update both cards at once.

**Comment by ghost** [comment]
> And you can group the cards under a higher-level heading:

But that way there would be 3 separate cards and they are grouped only in the org file, not in the anki, or am I missing something?

Another type of card I was thinking about was a card with all needed info on the front, ie definition, examples, maybe a picture, and the word, with optional extras like audio, on the back. It could be done by reversing the card, so I just haven't thought this through. Still, adding "extra" in heading will be clunky.

**Comment by eyeinsky** [comment]
> But that way there would be 3 separate cards and they are grouped only in the org file, not in the anki, or am I missing something?

Right, but as hwiorn said above, "Anki doesn't support overlapping clozes directly. The recommended way is that makes new text fields for every cloze."

I don't have anything against subheadings in principle, so if there is any org syntax examples and how that would translate to Anki cards then we could discuss it further.

**Comment by ghost** [comment]
> Right, but as hwiorn said above, "Anki doesn't support overlapping clozes directly. The recommended way is that makes new text fields for every cloze."

Cloze overlapper works with javascript, it wasn't updated for latest anki, so I can't test it. 

LPCG works by exporting separate cards, but, unlike one-liners, it provides a context needed for prose and lists. This method can be applied without subheadings by programmatically cutting list or prose using some org properties to set options, the same way [LPCG does it](https://ankilpcg.readthedocs.io/en/latest/importing.html).

Anyway, I stumbled upon [org-mode-incremental-reading](https://github.com/vascoferreira25/org-mode-incremental-reading) which uses special blocks instead of headings, maybe that's a better idea for defining extra fields - no need for subheadings and can be used for incremental reading. And maybe this method can be adapted for use as a delimiter https://github.com/eyeinsky/org-anki/issues/8.

**Comment by eyeinsky** [comment]
I've merged deck import to master (093ba9d2ab146d3548604205011c41c145a1f590, version bumped to v0.0.11).

It works via `org-anki-import-deck` command. It:
- prompts for deck name (when called interactively)
- runs pandoc, which is required to be available, on note's front and back to convert html to org, then
- writes output as org entries into current buffer

For me it has worked best to switch to a temporary buffer, import and edit there before adding it to any final location.
